trillian/ctfe/storage/storage.go

```diff
@@ -0,0 +1,17 @@
+package storage
+
+import (
+	"context"
+)
+
+// NonLeafCertificateChainStorage is an interface which allows CTFE binaries to use different storage implementations for non-leaf certificate chains.
+type NonLeafCertificateChainStorage interface {
+	// FindAll returns all key-value pairs of non-leaf certificate chains.
+	FindAll(ctx context.Context) (map[string][]byte, error)
```

Comment:
This could return a lot of data. Is this used enough to warrant being on the primary interface?

Reply:
Good question. I do not expect there will be a lot of issuance chains stored in the database. The number of chains could go up when time goes, then it is indeed possible to exceed the Go instance memory limit. This method is used to preload the chains from database into the cache.

Switching to `FindTop`/`FindLimit` or streaming data via Go channel?

The current MySQL schema design does not support selecting records with a predictable order. 
https://github.com/roger2hk/certificate-transparency-go/blob/a1aeb9c78da3fe10c24284597e1949080d0940eb/trillian/ctfe/storage/mysql/schema.sql#L18-L24

Reply:
From the data that you've looked at in the past, it shouldn't really be a lot of data if I remember correctly, so that shouldn't be a problem. In practice, there should be default max cache size and number of entries to protect CTFE's memory.

Reply:
Removing the `FindAll` method as the performance difference is minimal with/without the cache preload.